src/blocks/content-helper/current-post-details/provider.ts

```diff
@@ -0,0 +1,205 @@
+/**
+ * External dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import { select } from '@wordpress/data';
+import { addQueryArgs } from '@wordpress/url';
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * Internal dependencies
+ */
+import {
+	PostPerformanceData,
+	PostPerformanceReferrerData,
+} from './post-performance-data';
+
+/**
+ * Specifies the form of the response returned by the `/analytics/post/detail`
+ * WordPress REST API endpoint.
+ */
+ interface analyticsApiResponse {
+	error?: object;
+	data: PostPerformanceData[];
+}
+
+/**
+ * Specifies the form of the response returned by the `/referrers/post/detail`
+ * WordPress REST API endpoint.
+ */
+interface referrersApiResponse {
+	error?: object;
+	data: PostPerformanceReferrerData;
+}
+
+/**
+ * Provides current post details data for use in other components.
+ */
+class CurrentPostDetailsProvider {
+	private dataPeriodDays: number;
+	private dataPeriodStart: string;
+	private dataPeriodEnd: string;
+
+	/**
+	 * Constructor.
+	 */
+	constructor() {
+		// Return data for the last 7 days (today included).
+		this.setDataPeriod( 7 );
+	}
+
+	/**
+	 * Returns details about the post that is currently being edited within the
+	 * WordPress Block Editor.
+	 *
+	 * @return {Promise<PostPerformanceData>} The current post's details.
+	 */
+	public async getCurrentPostDetails(): Promise<PostPerformanceData> {
+		const editor = select( 'core/editor' );
+
+		// We cannot show data for non-published posts.
+		if ( false === editor.isCurrentPostPublished() ) {
+			return Promise.reject(
+				__( 'This post is not published, so its details are unavailable.', 'wp-parsely' )
+			);
+		}
+
+		// Get post URL.
+		const postUrl = editor.getPermalink();
+
+		// Fetch all needed results using our WordPress endpoints.
+		let performanceData, referrerData;
+		try {
+			performanceData = await this.fetchPerformanceDataFromWpEndpoint( postUrl );
+			referrerData = await this.fetchReferrerDataFromWpEndpoint( postUrl, performanceData.views );
+		} catch ( error ) {
+			return Promise.reject( error );
+		}
+
+		const period = { start: this.dataPeriodStart, end: this.dataPeriodEnd, days: this.dataPeriodDays };
+		return { ...performanceData, referrers: referrerData, period };
+	}
+
+	/**
+	 * Fetches the performance data for the current post from the WordPress REST
+	 * API.
+	 *
+	 * @param {string} postUrl
+	 * @return {Promise<PostPerformanceData> } The current post's details.
+	 */
+	private async fetchPerformanceDataFromWpEndpoint( postUrl: string ): Promise<PostPerformanceData> {
+		let response;
+
+		try {
+			response = await apiFetch( {
```

Comment:
Not sure, but does `apiFetch` accept a type parameter? I.e., maybe you could do `apiFetch<ReferrersApiResponse>()` instead of using `as`